### PR TITLE
DOC: manually placing images example

### PR DESCRIPTION
**Pull request by jklymak**
New example showing a couple of methods for dealing with images that you want to keep the same _reelative_ size.

### CODE REVIEW

**Comment by jklymak** [comment]
See rendered: https://output.circle-artifacts.com/output/job/6713e1a5-bef5-4df5-9311-82dfe81f5c04/artifacts/0/doc/build/html/gallery/images_contours_and_fields/image_exact_placement.html#sphx-glr-gallery-images-contours-and-fields-image-exact-placement-py

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+axs[0].imshow(A, vmin=-1, vmax=1)
+axs[1].imshow(B, vmin=-1, vmax=1)
+rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
+axs[0].add_patch(rect)
+
+# %%
+# Note that both images are rendered at a 1:1 ratio, but are made to look almost the
+# same width, despite image B being smaller than image A.
+#
+# We can control the relative sizes using the *width_ratios* argument _if_ the images
```
Comment (rcomer) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:55]: The underscores on the “if” are showing in the rendered ouput.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+"""
+=========================================
+Placing images, preserving relative sizes
+=========================================
+
+By default Matplotlib resamples images created with `~.Axes.imshow` to
+fit inside the parent ~.Axes`.  This can mean that images that have very
+different original sizes can appear similar in size.
+
+Sometimes, however,  it is desirable to keep the mages the same relative size, or
```
Comment (rcomer) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:10]:
```suggestion
Sometimes, however,  it is desirable to keep the images the same relative size, or
```

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+axs[0].imshow(A, vmin=-1, vmax=1)
+rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
+axs[0].add_patch(rect)
+axs[0].set_anchor('NW')
+axs[1].imshow(B, vmin=-1, vmax=1)
+axs[1].set_anchor('NW')
+
+# %%
+# Note that this procedure still leaves large white spaces (that can be trimmed in a
+# final product by ``bbox_inches="tight"`), and is not very general.  For instance, if
```
Comment (rcomer) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:82]:
```suggestion
# final product by ``bbox_inches="tight"``), and is not very general.  For instance, if
```
I also wonder if it’s worth making this a link to `savefig` so that it’s _really_ obvious what you’re referring to.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+# the axes had been arrayed vertically instead of horizontally, setting the height
```
Comment (rcomer) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:83]:
```suggestion
# the axes had been arranged vertically instead of horizontally, setting the height
```
?

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+# ================
+#
+# We can manually place axes when they are created by passing a position to
+# `~.Figure.add_axes`.  This position takes the form ``[left bottom width height]`` and
+# is in units that are a fraction of the figure width and height. Here we decide how
+# large to make the axes based on the size of the images, and add a small buffer of
+# 0.35 inches.  We do all this at 100 dpi.
+
+dpi = 100  # 100 pixels is one inch
+buffer = 0.35 * dpi  # inches
```
Comment (rcomer) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:97]: This comment feels ambiguous: the 0.35 is inches but `buffer` is in pixels.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+ax.add_patch(rect)
+
+ax = fig.add_axes([posB[0] / figw, posB[1] / figh, posB[2] / figw, posB[3] / figh])
+ax.imshow(B, vmin=-1, vmax=1)
+
+# %%
+# Inspection of the image will show that it is exactly 3* 35 + 300 + 200 = 605 pixels
+# wide, and 2 * 35 + 100 = 170 pixels high (or twice that if the 2x version is
+# used instead).  The images should be rendered with exactly 1 pixel per data point
+# (or two, if 2x).
```
Comment (rcomer) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:127]:
```suggestion
# (or four, if 2x).
```
?

**Review by rcomer — COMMENTED** [review]
This seems good to me (and also familiar… 😉).  Just a few minor comments.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py (hunk 6 above)
Comment (jklymak) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:127]: Sure.

**Comment by jklymak** [comment]
Thanks @rcomer - yes this came from the latest stack overflow to ask for this.  It is asked periodically, and I did not see that we had a good example to point to.

**Comment by jklymak** [comment]
Newest version: https://output.circle-artifacts.com/output/job/bb0ee1af-53af-43f3-a1d1-7b754a2c95ea/artifacts/0/doc/build/html/gallery/images_contours_and_fields/image_exact_placement.html#sphx-glr-gallery-images-contours-and-fields-image-exact-placement-py

**Review by timhoffm — COMMENTED** [review]
I suggest to add `# sphinx_gallery_thumbnail_number = -1` so that the last image will be used as thumbnail.

**Comment by jklymak** [comment]
> I suggest to add `# sphinx_gallery_thumbnail_number = -1` so that the last image will be used as thumbnail.

Yes, good idea - I also toyed with the idea of making the default figure size more appropriate, but I think it's reasonable to use the default so folks can see how it can "go wrong".

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+fig, axs = plt.subplots(1, 2, facecolor='aliceblue')
+
+axs[0].imshow(A, vmin=-1, vmax=1)
+axs[1].imshow(B, vmin=-1, vmax=1)
+rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
+axs[0].add_patch(rect)
+
+# %%
+# Note that both images are rendered at a 1:1 ratio, but are made to look almost the
+# same width, despite image B being smaller than image A.
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:56]:
```suggestion
# Note that both images have an aspect ratio of 1 (i.e. pixels are square), but
# pixels sizes differ because the images are scaled to the same width.
```

Optional suggestion: Use significantly lower resolution so that one can see the individual pixels.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+X, Y = np.meshgrid(x, y)
+R = np.sqrt(X**2 + Y**2)
+f0 = 5
+k = 100
+a = np.sin(np.pi * 2 * (f0 * R + k * R**2 / 2))
+A = a[:100, :300]
+B = A[:40, :200]
+
+# plot with default axes handling:
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:46]:
```suggestion
# default layout: both axes have the same size
```

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+even to make the images keep exactly the same pixels as the original data.
+Matplotlib does not automatically make either of these things happen,
+but it is possible with some manual manipulation.
+
+Preserving relative sizes
+=========================
+
+If the size of the images are amenable, we can preserve the relative sizes of two
+images by using either the *width_ratio* or *height_ratio* of the subplots.  Which
+one you use depends on the shape of the image and the size of the figure.
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:20]: I suggest to move this below the "default layout" plot (to line 58).

Train of thought:
- Problem statement
- Example illustration of the problem
- General solution
- Specific solution example

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+# helped because the axes are wider than they are tall. For more complicated
+# situations it is necessary to place the axes manually.
+#
+# Manual placement
+# ================
+#
+# We can manually place axes when they are created by passing a position to
+# `~.Figure.add_axes`.  This position takes the form ``[left bottom width height]`` and
+# is in units that are a fraction of the figure width and height. Here we decide how
+# large to make the axes based on the size of the images, and add a small buffer of
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:97]: Terminology: Would *margin* be a better name than *buffer*?

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+fig = plt.figure(figsize=(figw / dpi, figh / dpi), facecolor='aliceblue')
+
+ax = fig.add_axes([posA[0] / figw, posA[1] / figh, posA[2] / figw, posA[3] / figh])
+ax.imshow(A, vmin=-1, vmax=1)
+rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
+ax.add_patch(rect)
+
+ax = fig.add_axes([posB[0] / figw, posB[1] / figh, posB[2] / figw, posB[3] / figh])
+ax.imshow(B, vmin=-1, vmax=1)
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:125]: Suggestion structuring to make the calculations more understandable:
1. Calculate desired figure height and width in pixel
2. Create figure (needs calculation of figure size in inches)
3. Add axes (needs calculation of dimensions in figure coordinates) 

```suggestion
dpi = 100  # 100 pixels is one inch
margin_px = 0.35 * dpi  # in pixels
ny, nx = np.shape(A)
fig_height_px = ny + 2 * margin
fig_width_px = nx + 2 * margin

# calculate figsize in inches
figsize = (fig_width_px / dpi, fig_height_px / dpi)
fig = plt.figure(figsize=figsize, facecolor='aliceblue')

# add axes in figure coordinates
ax1 = fig.add_axes([
    margin_px / fig_width_px,   # left
    margin_px / fig_height_px,  # bottom
    nx / fig_width_px,          # width
    ny / fig_height_px          # height
])
ax2 = fig.add_axes([
    margin_px / fig_width_px,   # left
    (fig_height_px - margin_px - ny) / fig_height_px,  # bottom
    nx / fig_width_px,          # width
    ny / fig_height_px          # height
])

ax1.imshow(A, vmin=-1, vmax=1)
rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
ax.add_patch(rect)
ax2.imshow(B, vmin=-1, vmax=1)
```

File: galleries/examples/images_contours_and_fields/image_exact_placement.py (hunk 7 above)
Comment (jklymak) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:56]: Can we do that with sphinx-gallery?

File: galleries/examples/images_contours_and_fields/image_exact_placement.py (hunk 11 above)
Comment (jklymak) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:125]: A and B are purposefully different widths, so I don't think this is correct.

Comment (timhoffm) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:125]: You  are right. The numbers / expressions are not correct. Still the idea (which possibly can be communicated more explicitly) is that you can place everything at exact pixel positions. And to do that, you have to follow the three steps above.

Comment (timhoffm) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:125]: Possibly now correct. - I don't have an interpreter at hand, so can't check, but I hope you get the idea.

Comment (jklymak) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:125]: I do get the idea, but not convinced pre-calculating the figure dimensions buys you anything.  Also, imagine we were to add a third or fourth image in there, we would need to change the code in two places.

Just for context, I originally wrote as a for-loop to allow flexibility, but backed off on that as perhaps too magical.

Comment (timhoffm) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:125]: > Just for context, I originally wrote as a for-loop to allow flexibility, but backed off on that as perhaps too magical.

👍 Examples must show the pattern, not necessarily a generic solution if that makes the code harder to understand.

---

I had difficulty following your code. Reflecting on it, the reasons were
1. lines 100-115 define a lot of variables, which are sometimes
  a) tersely named (`figh`)
   b) cryptic data structures `posB = [left, figh - buffer - ny, nx, ny]`
   c) a mixture of definitions for later use (e.g. `figh`) and temporary local variables that are reassigned (`left`, `nx`, `ny´)
2. It's unclear that these are all in pixels
3. Transformation calculations to inches (for figsize) and figure units (for add_axes) are hard to understand, because both the input units (see 2.) and the resulting units are not clear.

I tried a more minimal refactor of your code to alleviate these points, but I ended up with almost the same code as I've proposed above 🤷‍♂️.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py (hunk 7 above)
Comment (timhoffm) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:56]: What I meant has nothing to do with sphinx-gallery: Use N=50 or 100 instead of 450, and equally scale down your image slices. You may need to adjust you formula for `a` to have a reasonably slowly varying function so that it does not change too much between pixels on the coarse pixel grid.

Comment (jklymak) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:56]: So make the images small when I do the explicit placement below? Certainly they could be smaller, but maybe not too much smaller?

**Comment by jklymak** [comment]
The doc build is failing with 

```
File "/home/circleci/project/galleries/examples/units/radian_demo.py", line 26, in <module>
        axs[0].plot(x, cos(x), xunits=radians)
                       ^^^^^^
      File "/home/circleci/project/galleries/examples/units/basic_units.py", line 382, in cos
        return [math.cos(val.convert_to(radians).get_value()) for val in x]
                         ^^^^^^^^^^^^^^
    AttributeError: 'numpy.float64' object has no attribute 'convert_to'
```

**Comment by timhoffm** [comment]
Please rebase. #28779 fixes the doc build.

**Comment by timhoffm** [comment]
> I also toyed with the idea of making the default figure size more appropriate, but I think it's reasonable to use the default so folks can see how it can "go wrong".

I'm +1 on adjusting the figure size right away for all figures. Adjusting the figsize is something one would always do if the default aspect is not a good fit for the content - e.g. if you have two same-size images of your data side-by-side.

Just looking at the figures, one could get the impression, that it's a feature of the "manual" approach that you don't have the large figure whitespace. To take this effect out, *all* figures should have the same (or similar) size. While not being a feature to be explicitly highlighted, figsize is a "standard adjustment" that one would *just* do, as we do in other examples like https://matplotlib.org/stable/gallery/lines_bars_and_markers/categorical_variables.html.

**Comment by jklymak** [comment]
I'm not sure I agree with that.  Of course people _should_ change the figure size, but there is endless confusion about fixed ratio images and why there is so much white space.  If we show the white space, I think on the balance we are helping folks understand this is not something matplotlib will do for them automagically.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+axs[1].imshow(B, vmin=-1, vmax=1)
+axs[1].set_anchor('NW')
+
+# %%
+# Note that this procedure still leaves large white spaces (that can be trimmed
+# in a final product by ``bbox_inches="tight"`` in `~.Figure.savefig`), and is
+# not very general.  For instance, if the axes had been arranged vertically
+# instead of horizontally, setting the height aspect ratio would not have
+# helped because the axes are wider than they are tall. For more complicated
+# situations it is necessary to place the axes manually.
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:96]: I personally think, that choosing an appropriate figure size right away is better than cutting afterwards using``bbox_inches="tight"``. As mentioned above the appropriate figure size has dicdactic advantages here as well and that alleviates the need to discuss cutting. You can then simply leave this paragraph out.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+#
+# Manual placement
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:98]: I'm a bit torn on the motivation for this: You can get away without by using an appropriate figsize (which is much simpler).

The advantages I see is
1. pixel-perfect positioning (if you need that)
2. in this case 1:1 mapping from data points to visual pixels.

Maybe these should be highlighted more.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py (hunk 12 above)
Comment (jklymak) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:96]: yeah, I disagree that there is a didactic advantage to manually setting the figsize in this case.  Just the opposite: "here we set the figsize based on our trial and error to make the rest of this particular plot look good" does't teach much.  Whereas, here is the default figsize - note how there is white space by default that you will need to remove manually is probably what most people are going to see to start.  

I mean I guess a default guess that the aspect ratio of the figure will be similar to the aspect ratio of the original matrices is a middle ground.   But again, I will re-iterate that there are _many_ questions of stack overflow where the user seems confused about how Matplotlib handles fixed-aspect axes.  I think there is value in showing what happens by default, and chose this route explicitly because of that.

Comment (timhoffm) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:96]: I agree that figure figsize is a topic that needs explanation. I'm just not sure whether it should be part of *this* example. The focus is on placing and sizing images. Adding the figsize topic into the mix may be too distracting. But if you think it's in scope, I suggest to discuss it right at the start after the all-default figure. It feels awkward to discuss and illustrate placement on a series of ill-sized figures.

**Comment by jklymak** [comment]
@timhoffm perhaps this is worse, but this change I _don't_ set the figsize for the first "default" plot, but then _do_ set it for the customized plots, with a bit of explanation.   That doesn't make the narrative shorter, but it does go through the steps in the way a user may go through them. 

https://output.circle-artifacts.com/output/job/d67643fd-551a-4486-8c8a-409ddbdef5ba/artifacts/0/doc/build/html/gallery/images_contours_and_fields/image_exact_placement.html#sphx-glr-gallery-images-contours-and-fields-image-exact-placement-py

**Comment by timhoffm** [comment]
I like this. It's clear without making too much fuss about the figsize topic.

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+# because the axes are wider than they are tall.
+#
+# Manual placement
+# ================
+#
+# We can manually place axes when they are created by passing a position to
+# `~.Figure.add_axes`.  This position takes the form ``[left bottom width
+# height]`` and is in units that are a fraction of the figure width and height.
+# Here we decide how large to make the axes based on the size of the images,
+# and add a small buffer of 0.35 inches.  We do all this at 100 dpi.
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:113]:
```suggestion
# Explicit placement
# ==================
# The above approach with adjusting ``figsize`` and ``width_ratios`` does
# not generalize well, because it needs manual parameter tuning, and
# possibly even code changes to using ``height_ratios`` instead of 
# ``width_ratios`` depending on the aspects and layout of the images.
#
# We can alternative calculate positions explicitly and place Axes at absolute
# coordinates using `~.Figure.add_axes`. This takes the position in the form
# ``[left bottom width height]`` and is in
# :ref:`figure coordinates <transforms_tutorial>`. In the following, we
# determine figure size and Axes positions so that one image data point
# is rendered exactly to one figure pixel.
```

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
+"""
+=========================================
+Placing images, preserving relative sizes
+=========================================
+
+By default Matplotlib resamples images created with `~.Axes.imshow` to
+fit inside the parent `~.axes.Axes`.  This can mean that images that have very
+different original sizes can end up appearing similar in size.
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:8]: Optional: Slightly more concise wording:
```suggestion
fit inside the parent `~.axes.Axes`.  As a result, images with very different
original sizes may end up appearing similar in size.
```

File: galleries/examples/images_contours_and_fields/image_exact_placement.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+Sometimes, however,  it is desirable to keep the images the same relative size, or
+even to make the images keep exactly the same pixels as the original data.
+Matplotlib does not automatically make either of these things happen,
+but it is possible with some manual manipulation.
```
Comment (timhoffm) [inline galleries/examples/images_contours_and_fields/image_exact_placement.py:13]: Optional: Slightly more concise wording:
```suggestion
This example shows how to keep the images the same relative size, or
how to make the images keep exactly the same pixels as the original data.
```

File: galleries/examples/images_contours_and_fields/image_exact_placement.py (hunk 8 above)
Comment (timhoffm) [reply galleries/examples/images_contours_and_fields/image_exact_placement.py:42]: Ping here: Have you overlooked this? If you don't want to take it, please give a short feedback anyway.

**Review by timhoffm — APPROVED** [review]
Take or leave the comments. May self-merge afterwards.
